cocotb/share/lib/embed/gpi_embed.c

```diff
@@ -243,121 +268,133 @@ int embed_sim_init(gpi_sim_info_t *info)
         goto cleanup;
 
     // Obtain the loggpi logger object
-    simlog_obj = PyObject_GetAttrString(cocotb_module, "loggpi");
+    simlog_obj = PyObject_GetAttrString(cocotb_module, "loggpi");       // New reference
 
     if (simlog_obj == NULL) {
         PyErr_Print();
         LOG_ERROR("Failed to get simlog object from loggpi\n");
+        goto cleanup;
     }
 
-    simlog_func = PyObject_GetAttrString(simlog_obj, "_logFromC");
-    if (simlog_func == NULL) {
+    // Obtain the function to use when logging from C code
+    log_func = PyObject_GetAttrString(simlog_obj, "_logFromC");         // New reference
+    if (log_func == NULL) {
         PyErr_Print();
         LOG_ERROR("Failed to get the _logFromC method");
         goto cleanup;
     }
 
-    if (!PyCallable_Check(simlog_func)) {
+    if (!PyCallable_Check(log_func)) {
         PyErr_Print();
         LOG_ERROR("_logFromC is not callable");
+        Py_DecRef(log_func);
```

Comment:
What's the rationale for using the non-macro versions? I've only ever seen the macro versions used  in other codebases.

Reply:
From the documentation:
"The following functions are for runtime dynamic embedding of Python: Py_IncRef(PyObject *o), Py_DecRef(PyObject *o). They are simply exported function versions of Py_XINCREF() and Py_XDECREF(), respectively."

I'm not strongly attached. I could use the macros and be explicit about Py_DECREF vs Py_XDECREF (implied use under ok label).

Reply:
> for runtime dynamic embedding of Python

My understanding is that this is for when you are trying to pull `Py_IncRef` from a dll, and do not have any header files available. Given that's not our use case, I think we should use the macro versions.

On this particular line, you can use `Py_DECREF`, since you are sure that it is not null.

Reply:
Changed to macro versions.
I used Py_XDECREF in a few places where appropriate.